src/row-data-utils.ts

```diff
@@ -193,14 +193,23 @@ class RowDataUtils {
       properties.reqOpts
     );
     properties.requestData.data = {};
+    const passThroughCallback = (
+      err: ServiceError | null,
+      apiResponse?: google.bigtable.v2.IReadModifyWriteRowResponse
+    ) => {
+      if (err && err.message === 'The client has already been closed.') {
+        err.code = 1; // CANCELLED
```

Comment:
Attaching a code here is required for the test to pass, but maybe we should just modify the test runner so that it checks for a message instead of an error code and leave this method unchanged.

Reply:
I don't like the idea of reading an error string to change the exception type, it seems hacky. Can we make it raise the proper error originally, instead of overwriting it?

Reply:
I don't like the magic string matching either. The problem is that the error [gets raised in the Gapic layer](https://github.com/googleapis/nodejs-bigtable/blob/86f99f59ac530f2ead89ce817f5e20b47d28728d/src/v2/bigtable_client.ts#L319) so we can't make the change where it gets raised originally without regenerating gapic.

Another alternative is to add another check to the if condition `&& !err.code` to only insert a code if the code is missing. What do you think?

Reply:
Something seems a bit off here. My understanding of [part 4b](https://github.com/googleapis/cloud-bigtable-clients-test/blob/595412bc54322d9e21b7df5edf33e732561f5786/tests/readmodifywriterow_test.go#L333) is it's meant to detect rpcs cancelled in-flight, so it should be the backend raising the error here, not the gax layer. The [code you posted](https://github.com/googleapis/nodejs-bigtable/blob/86f99f59ac530f2ead89ce817f5e20b47d28728d/src/v2/bigtable_client.ts#L319) seems to be for rejecting rpcs started after the client is already closed, which would be verified by [part 4c of the test](https://github.com/googleapis/cloud-bigtable-clients-test/blob/595412bc54322d9e21b7df5edf33e732561f5786/tests/readmodifywriterow_test.go#L349)

This is a bit of a complicated situation though, so let me know if I'm missing something. I think I'll need to dive a bit deeper to really give a recommendation here

Reply:
I'm actually a bit unsure why `Cancelled` is even acceptable for this test? Looking at the test, I'd expect all the [batch 1](https://github.com/googleapis/cloud-bigtable-clients-test/blob/595412bc54322d9e21b7df5edf33e732561f5786/tests/readmodifywriterow_test.go#L337) rpcs to be successful, since the client isn't closed until they are completed...

Reply:
all the batch 1 rpcs _are_ successful. The assertion statement that this change addresses is [here](https://github.com/googleapis/cloud-bigtable-clients-test/blob/595412bc54322d9e21b7df5edf33e732561f5786/tests/readmodifywriterow_test.go#L353). Looking into this further...

Reply:
[The assertion statement throwing the error](https://github.com/googleapis/cloud-bigtable-clients-test/blob/595412bc54322d9e21b7df5edf33e732561f5786/tests/readmodifywriterow_test.go#L353) throws an error if the code is 0 or if it is not provided. I am okay with using a different error code, but we must use a non-zero error code to address this assert statement. If you want to see why the code adjustment here is necessary, try running the conformance test and commenting out this code change.

What do you think we should do?

Reply:
The test proxy service returns a status to the runner.

It might make it easier to see with a screenshot of what happens with these changes:

![image](https://github.com/user-attachments/assets/26d70b4a-eab8-44c1-9b4d-90513c754deb)

If you set a breakpoint just before the test proxy sends back the status, you can see that the error code that will be sent back is 1. Without these changes, `e` has a `stack` and a `message`, but not a code. So it looks like the close client error from gapic is not providing an error code.

Reply:
Yeah, my confusion was over why a method that returns an exception was being passed as successful, but I think I understand now. The gapic layer is using `this._gaxModule.GoogleError` withot a grpc code, since the error is client side, and not from a grpc channel. And the test proxy is interpreting no grpc code as `0`, or `OK`

So to address this, I think you should change line 45 in your screenshot to something like `code: e.code ?? grpc.status.UNKNOWN`. Currently, the test proxy is reporting that exceptions without a grpc status code are OK, but that's not what the client is actually telling the proxy. The client had an error, but the proxy is covering it up. We need to fix that

Reply:
Ok. We need to make that particular change for all the services then.

Reply:
Ok, sounds good